### ISSUE TITLE
Decrease memory consumption of pre-target buffer

### ISSUE THREAD

**Issue by bluerogue251** [bluerogue251/DBSubsetter#33]
Currently one of the major consumers of heap memory is the Pre-Target-DB-Insert Buffer. This is a queue that holds the complete row data for any rows that are waiting to be inserted to the target DB. Because the target DB inserts are often a bottleneck in the subsetting process, this queue can grow quite large and take up a lot of memory.

We currently manage this by allowing users to specify a maximum size for this buffer with the `--preTargetBufferSize` option. While this prevents `OutOfMemory` and `GarbageCollectionOverhead` errors, it is still basically just a bandaid. A better solution would either make this buffer fundamentally take up less heap space so that its size could grow as much as needed, or it would quicken the pace of target DB inserts so that having a large buffer was no longer necessary.

A few brainstorming ideas, (more discussion and thought is definitely needed on all of these):
  (1) Keep the pre-target buffer in Chronicle-Queue instead of on heap. This would solve the memory consumption problem, however since the number of possible data types is very large, it might be more difficult to write serialization/deserialization logic.
  (2) Improve performance of target-DB inserts by using Bulk Copy functionality such as `copy from stdin`, etc. This is probably an overall win, but one con is that the insert logic would then have to be vendor-specific. Serialization might still be an issue if certain vendors turn out only to accept binary or CSV input.
  (3) Instead of keeping all the columns of the rows in memory, we could just keep the primary keys in memory. Then, for a target insert, we would query the full row data out of the origin DB and stream it into the target db. We could use bulk copy APIs here as well. This way, most of the row data never enters java heap space. Downsides to this include potentially double the amount of roundtrips to the origin database, but it might turn out to be well worth it. This might also help us with https://github.com/bluerogue251/DBSubsetter/issues/17 if we are able to stream binary or csv data directly from origin db to target db, without needing to convert advanced data types into java equivalents in the middle.
  (4) Do some batching of target inserts. For instance, if we are about to insert 50,000 rows into a table, try to do that in as few SQL insert statements as possible. Also, we would only take any SQL actions if a certain amount of data had been built up -- for instance, we wouldn't make any calls to the target DB until we had at least 1,000 rows to insert at once. This would improve insert performance and in that way reduce the amount of data in this buffer for certain use cases, if this were no longer the bottleneck. In other cases, it would only make the problem more severe by holding more in memory... Batching may be of use even if we were to do (3) -- in this case the batching would work on the primary keys rather than on the entire row data...

**Comment by bluerogue251** [comment]
The current plan for this is to do it in several steps in a long-standing integration branch

1) Enable Target DB Inserts by Primary Key Alone
2) Convert Target DB Buffer from in-memory containing full row data to off-heap chronicle queue containing primary key values only (single row at a time will be inserted into the target DB -- performance at this point will suffer greatly)
3) Enable batching of Target DB Inserts to allow for more rows per insert statement, on average (helping to gain back reasonable performance)

---- Here we do load testing and in theory we should see performance gains ----
---- If so, we can merge the long-standing integration branch ----

Then after the long-standing integration branch is merged we can follow up with this cleanup:

4) Remove all non-primary-key and non-foreign-key row data from origin DB key-calculation queries, as it is no longer used

**Comment by bluerogue251** [comment]
Went through with the plan. We now use chronicle queue as an on-disk buffer of rows which we need to insert into the target database. We only store primary key values for these rows on disk -- we do not store all the columns from each row. We dequeue in batches so that we can improve insert performance by inserting up to one or two thousand rows at a time. On every dequeue, we have to go to the origin database and use the primary key value to fetch the full row data for each row. We then insert it into the target database.

It was non-trivial to do a stress test to ensure that Chronicle Queue can handle hundreds or thousands of queues at once, so we skipped this test and will simply cross our fingers and hope for the best in that regard.

We have also updated all of our key-calculation origin db queries to only select columns related to primary or foreign keys (as opposed to how it used to be, where all columns from the table were selected).

Did the load testing and did not see much gained or lost performance wise. However, it still feels safer from a memory management perspective, as well as more flexible in general, to have the target buffer live on disk rather than in memory, so we will stick with this for the time being.
